ignite/distributed/auto.py

```diff
@@ -0,0 +1,101 @@
+import sys
+import functools
+import torch
+
+from torch.utils.data import DataLoader
+from torch.utils.data.distributed import DistributedSampler
+
+from ignite.distributed.utils import get_world_size, get_backend, get_local_rank, get_rank, get_device
+from ignite.distributed.utils import initialize, finalize
+
+
+def auto_batch_size(batch_size: int):
+    return batch_size // get_world_size()
+
+
+def auto_num_workers(num_workers: int):
+    if get_backend() == "nccl":
+        ngpus = torch.cuda.device_count()
+        return (num_workers + ngpus - 1) // ngpus
+    else:
+        return num_workers
+
+
+def auto_dataloader(dataset, batch_size=1, num_workers=0, pin_memory=False, drop_last=False, shuffle=False):
+    # TODO tpu !!
+    if get_world_size() > 1:
+        return DataLoader(
+            dataset,
+            sampler=DistributedSampler(dataset, shuffle=shuffle),
+            batch_size=auto_batch_size(batch_size),
+            num_workers=auto_num_workers(num_workers),
+            pin_memory=pin_memory,
+            drop_last=drop_last,
+        )
+    else:
+        return DataLoader(
+            dataset,
+            batch_size=batch_size,
+            num_workers=num_workers,
+            pin_memory=pin_memory,
+            drop_last=drop_last,
+            shuffle=shuffle,
+        )
+
+
+def auto_model(model):
+    # TODO tpu !!
+    # distributed data parallel model
+    if get_world_size() > 1:
+        if get_backend() == "nccl":
+            local_rank = get_local_rank()
+            return torch.nn.parallel.DistributedDataParallel(model, device_ids=[local_rank,], output_device=local_rank)
+        elif get_backend() == "gloo":
+            return torch.nn.parallel.DistributedDataParallel(model)
+        else:
+            raise ValueError("autodistributed_model can not handle backend {}".format(get_backend()))
+    # not distributed but multiple GPUs reachable so data parallel model
+    elif torch.cuda.device_count() > 0 and get_device() == "cuda":
+        return torch.nn.parallel.DataParallel(model)
+    # sequential model
+    else:
+        return model
+
+
+def auto_distributed(backend=None, device=None):
```

Comment:
For XLA we need to spawn decorated user function on `nprocs` like here:
https://github.com/pytorch/ignite/pull/988/files#diff-7ae8bdbd21c0e1013d012c1cb9b52457R35

Reply:
I will check

Reply:
Ok I see. That is related to `spawn` mode that I would like provide.